README.md
# Domoticz Internal Script Viewer

# Objectives
* To view offline the Domoticz internal scripts (focus on dzVents scripts), which are created using the web ui event editor.
* To filter by script type.
* To highlight syntax depending selected script type, i.e. XML, Python and for dzVents some Lua constants, keywords, objects.
* To export the scripts to the clipboard, text file or HTML file.
* To develop the application in Lazarus running on a Windows 10 device.

**Developed for personal use only under the GNU GENERAL PUBLIC LICENSE Version 3.**

![domisv-m](https://user-images.githubusercontent.com/47274144/117531946-51d1af00-afe5-11eb-99b0-dc5c22441f3d.png)

![domisv-s](https://user-images.githubusercontent.com/47274144/117531947-526a4580-afe5-11eb-8b73-e43b1cdbff94.png)

## Solution
On the Domoticz Server (Raspberry Pi 3B+) the folder /home/pi/domoticz is defined as shared folder(samba).
In Windows a network drive is defined (Z: as an example) using the Domoticz shared folder.
The viewer copies, from drive Z:, the Domoticz database, domoticz.db, to the local folder in which the viewer is running.
From the database, table EventMaster, the all records with fields Name, Interpreter, XMLStatement and Status are selected and displayed in a grid.
Selecting a script shows the script code (field XMLStatement) in a textarea.

## Hints
The application enables to download the Domoticz database from the Domoticz server. It is therfore also an option to backup the Domoticz database.

## Software
* Raspberry Pi Raspian Linux 4.14.79-v7+ #1159.
* Domoticz Home Automation System 2021.1.
* Developed with Lazarus v2.0.12 and FPC 3.2.0 - Tested under Windows 10, Ubuntu 20.04.
* Note: Under Ubuntu ensure to install: sudo apt-get install libsqlite3-dev.

## Install
Create and connect to the shared folder (see next how to).
Install the application in a folder of choice and start ***domoticzinternalscriptviewer.exe***.

#### Domoticz Server Shared Folder
The Raspberry Pi shared folder uses samba and is configured with:
```sudo nano /etc/samba/smb.conf```
Content:
```
[DoProDomoticz]
Comment = Raspberry Pi Domoticz Production folder
Path = /home/pi/domoticz
Browseable = yes
Writeable = Yes
only guest = no
create mask = 0777
directory mask = 0777
Public = yes
Guest ok = yes
```
After the samba configuration:
* Ensure to set the directory mask for the domoticz folder to 0777: sudo chmod 777 /home/pi/domoticz
* Restart samba: sudo /etc/init.d/samba restart
* Restart domoticz: sudo service domoticz.sh restart

#### Windows Network Drive Connect to Shared Folder
In Windows connect to the Domoticz server shared folder using the Windows Explorer > Connect Network:
Path for drive Z: is \\192.nnn.n.nn\DoProDomoticz
(DoProDomoticz is defined in the samba configuration)

After this prework, the domoticz database domoticz.db can be copied using CopyFile.
(simple solution using the standard lazarus fileutil functions).

## Settings
The viewer uses following settings(stored in domoticzinternalscriptviewer.ini):
### Section general
#### shareddrive
The Windows shared drive, i.e. Z:\ or Y:\domoticz\.
#### databasefile
Domoticz database file copied and stored locally in the application folder. Default=domoticz.db.
#### refreshatstart
Flag to copy the database file at application start. Default=0.
#### sqlquery
SQL SELECT statement to read the records from table EventMaster.
Default:
SELECT Name,Interpreter,XMLStatement,Status FROM EventMaster ORDER BY Interpreter,Name;
#### filter
Filter for selecting script entries. 
Default: Interpreter="dzVents"
Other filter examples: 
Select active dzVents scripts: Interpreter="dzVents" and Status=1

### Section synedit
#### gutter
Show the left gutter: 0=Not visible, 1=Visible.
Default: 0
#### fontsize
Editor font size.
Default: 10

The next settings are related to dzVents or Lua syntax highlighting.
Change as required - these are just try outs.
#### constants
Highlighting constants for dzVents.
Default: DATA,DEVICES,HTTPRESPONSES,LOGGING,TIMER
#### objects
Highlighting objects for dzVents.
Default: LOCAL
#### keywords
Highlighting keywords for dzVents.
Default: ACTIVE,CUSTOMEVENTS,DATA,DEVICES,DO,DUMP,ELSE,END,EXECUTE,FALSE,FILTER,FIND,FOR,FOREACH,FORMAT,FUNCTION,GROUPS,HELPERS,HTTPRESPONSES,IF,IN,LEVEL,LOG,LOGGING,MARKER,NEXT,NOTIFY,ON,REDUCE,RETURN,SCENES,SECURITY,SHELLCOMMANDRESPONSES,SWITCHOFF,SWITCHON,SYSTEM,THEN,TIME,TIMER,TRUE,VARIABLES

## Disclaimer
THIS APPLICATION IS PROVIDED BY THE AUTHOR “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, 
INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
POSSIBILITY OF SUCH DAMAGE.


CHANGELOG.md
# ChangeLog domoticz-internal-script-viewer

### v1.1.0 (Build 20210507)
* NEW: Menu Edit - Filter to select scripts, i.e. Interpreter = 'dzVents' or Status = 1 or combine with AND.
* NEW: Settings - Added Filter to select scripts
* NEW: Syntax highligher - XML, Python and for dzVents some Lua constants, keywords, objects. Can be enhanced in the ini file.
* NEW: Menu Tools - Set Editor Gutter & Font Size.
* UPD: Changed layout Main Form - Left pane: list of scripts, middle: splitter to resize the left pane, right pane: editor (SynEdit).
* UPD: Settings - If settings sql query is changed, the Domoticz database is reloaded.
* UPD: Compiled with Lazarus v2.0.12 and FPC 3.2.0 - Tested under Windows 10, Ubuntu 20.04.
* UPD: Minor improvements & enhancements.

### v1.0.3 (Build 2000124)
* UPD: Compiled with Lazarus v2.0.6 X64.
* UPD: SQLite dll X64 3.31.0.

### v1.0.2 (Build 20190417)
* UPD: Compiled with Lazarus v2.0.2.

### v1.0.1 (Build 20190301)
* NEW: Export all scripts file-by-file.

### v1.0.0 (Build 20190201)
* NEW: First version.


source/umisc.pas
unit umisc;
{
@name(umisc)
@shorttitle(Global  helper routines)
@author(Robert W.B. Linn)
@seealso(umain, readme.txt)
@created(see umain)
@lastmod(see umain)
}
{REGION NOTES}
{
More options in progress for a next version of Look4How.
}
{ENDREGION NOTES}

{$mode objfpc}{$H+}

interface

uses
 Classes, SysUtils,
 typinfo,
 Clipbrd,
 comctrls,
 strutils,
 Graphics,            // TColor
 Forms,               // Application.MessageBox
 LCLIntf,             // PredefinedClipboardFormat(
 LCLType              // MB_nnnnn
 ;

{REGION FUNCIONS}
function BoolToString(AState: Boolean): String;
function BoolToYN(AState: Boolean): String;
function BoolToText(AState: Boolean; ATrueText, AFalseText: String): String;
function TabToSpace(AString: String; ANumSpaces: Integer): String;

function ColorToHTML(const AColor: TColor): String;

procedure CopyToClipboard(AString: String);
function PasteFromClipboard: String;
procedure ListViewToClipboard(AListView: TListView);
function FileToClipboard(AFileName:String; ADelete:Boolean):Boolean;

procedure StringToStringList(const ADelimiter: Char; AInput: string; const AStrings: TStrings);
function ReplaceAllIgnoreCase(AString, AOldValue, ANewValue: String): String;

procedure ShowInfo(ACaption, AMessage: String);
procedure ShowWarning(ACaption, AMessage: String);
procedure ShowError(ACaption, AMessage: String);
{ENDREGION FUNCIONS}

implementation

{REGION CONVERSION}
// Convert a boolean True|False to text string True|False
// Return True|False
function BoolToString(AState: Boolean): String;
begin
  Result := 'True';
  if NOT AState then Result := 'False';
end;

// Convert a boolean True|False to text string Yes|No
// Return Yes|No
function BoolToYN(AState: Boolean): String;
begin
  Result := 'Yes';
  if NOT AState then Result := 'No';
end;

// Convert a boolean True|False to text string TrueText|FalseText
// Set the text in the parameter TrueText|FalseText
// Return TrueText|FalseText
function BoolToText(AState: Boolean; ATrueText, AFalseText: String): String;
begin
  Result := ATrueText;
  if NOT AState then Result := AFalseText;
end;

// Convert a string to a stringlist
// Example:
// vContent := vContent.Replace(LineEnding, #10, [rfReplaceAll]);
// StringToStringList(#10, vContent, Lines);
// Returns a StringList
procedure StringToStringList(const ADelimiter: Char; AInput: string; const AStrings: TStrings);
begin
 Assert(Assigned(AStrings)) ;
 AStrings.Clear;
 AStrings.StrictDelimiter := true;
 AStrings.Delimiter := ADelimiter;
 AStrings.DelimitedText := AInput;
end;

// Replace all occurencies ignoring case
// Return Converted string
function ReplaceAllIgnoreCase(AString, AOldValue, ANewValue: String): String;
begin
 Result := AString.Replace(AOldValue, ANewValue, [rfReplaceAll, rfIgnoreCase]);
end;

// Convert tab characters to spaces
// Return Converted string
function TabToSpace(AString: String; ANumSpaces: Integer): String;
begin
  Result := Tab2Space(AString, ANumSpaces);
end;

{ENDREGION CONVERSION}

{REGION COLORS}
// Convert TColor to HTML HEX string #RRGGBB
// Returns #RRGGBB
function ColorToHTML(const AColor: TColor): String;
begin
  Result := '#' + HexStr(ColorToRGB(AColor), 6);
end;
{ENDREGION COLORS}

{REGION CLIPBOARD}
// Copy text string to the clipboard.
// The clipboard content is cleared prior copy.
procedure CopyToClipboard(AString: String);
begin
  Clipboard.Clear;
  Clipboard.SetTextBuf(PChar(AString));
end;

// Paste text string from the clipboard.
// The clipboard content is checked prior pasting.
// Returns the clipboard content as a string else empty string
function PasteFromClipboard: String;
begin
  Result := '';
  if not Clipboard.HasFormat(PredefinedClipboardFormat(pcfText)) then
  begin
    Result := 'The clipboard content has no text format.';
    Exit;
  end;
  Result := Clipboard.AsText;
end;

procedure ListViewToClipboard(AListView: TListView);
var
  StringList: TStringList;
  RowNr: Integer;
  ListItem : TListItem;
  TempStr: String;
  i: Integer;
begin
  // Do nothing if no items
  if AListView.Items.Count = 0 then Exit;
  // Build the stringlist
  StringList := TStringList.Create;
  // Loop over the rwos of the listview
  for RowNr := 0 to AListView.Items.Count - 1 do
  begin
    // Get the listitem
    ListItem := AListView.Items[RowNr];
    // Build tempstring containing the caption and subitems
    TempStr := ListItem.Caption;
    for i := 0 to ListItem.SubItems.Count - 1 do
    begin
      TempStr := TempStr + ','+ ListItem.SubItems[i];
    end;
    // Add the tempstring to the list
    StringList.Add(TempStr);
  end;
  // Copy to the clipboard
  CopyToClipboard(StringList.Text);
end;

// Copy  content of a text file to the clipboard.
// Option to delete the file.
function FileToClipboard(AFileName:String; ADelete:Boolean):Boolean;
var
  Strings: TStringList;
begin
  Result := False;
  try
    Strings := TStringList.Create;
    Strings.LoadFromFile(AFileName);
    CopyToClipboard(Strings.Text);
    if ADelete then
    begin
      DeleteFile(AFileName);
    end;
    Result := True;
  except
    on E: Exception do
    begin
      // E.Message
      Exit;
    end;
  end;
end;
{ENDREGION CLIPBOARD}

{REGION MESSAGES}
procedure ShowInfo(ACaption, AMessage: String);
begin
  Application.MessageBox(PChar(AMessage), PChar(ACaption), MB_ICONINFORMATION)
end;

procedure ShowWarning(ACaption, AMessage: String);
begin
  Application.MessageBox(PChar(AMessage), PChar(ACaption), MB_ICONWARNING)
end;

procedure ShowError(ACaption, AMessage: String);
begin
  Application.MessageBox(PChar(AMessage), PChar(ACaption), MB_ICONERROR)
end;
{ENDREGION MESSAGES}

end.



source/umain.pas
unit umain;
{
@name(umain)
@shorttitle(Main unit of the project)
@author(Robert W.B. Linn)
@seealso(umain, readme.txt)
@created(20190201)
@lastmod(see uconstants)
@note(Under Ubuntu ensure to run sudo apt-get install libsqlite3-dev)
}
{$mode objfpc}{$H+}

{ TODO 1 -cToolbar : Change toolbarmain by panel with speedbuttons to allow checkboxes, comboboxes etc. }
{ TODO 1 -cApplication : Use Lazarus icons from folder c:\Prog\lazarus\images\. }
{ DONE 3 -cEditor : Option to show the synedit gutter. }
{ DONE 3 -cEditor : Option to set the font size of synedit. }
{ DONE 3 -cApplication : Add menustructure File,Edit,Help. }

interface

uses
 Classes, SysUtils, FileUtil,
 SynEdit, SynEditKeyCmds, SynHighlighterAny,
 SynHighlighterCpp, SynHighlighterXML, SynHighlighterPython,
 Forms, Controls,
 Graphics, Dialogs, ComCtrls, ActnList, DBGrids, ExtCtrls, Menus, IniFiles,
 // project units
 uconstants, umisc, usettings
 // ,udownload = ON HOLD
 , LCLType, DB, Grids;

type

  { TFormMain }

  TFormMain = class(TForm)
     ActionToolsEditorGutter: TAction;
     ActionToolsEditorFontSize: TAction;
     ActionEditSelectAll: TAction;
     ActionEditPaste: TAction;
     ActionEditCopy: TAction;
     ActionEditCut: TAction;
     ActionEditFilterReset: TAction;
     ActionEditFilter: TAction;
     ActionFileLoad: TAction;
     ActionFileExportEachScriptToFile: TAction;
     ActionEditCopyScript: TAction;
     ActionToolsSettings: TAction;
    ActionHelpAbout: TAction;
    ActionFileExportAllToHTML: TAction;
    ActionFileExportAllToText: TAction;
    ActionFileRefresh: TAction;
    ActionFileExit: TAction;
    ActionListMain: TActionList;
    DBGridDomoticz: TDBGrid;
    ImageListMain: TImageList;
    MenuFileExportSingleFile: TMenuItem;
    MenuEditCopyScript: TMenuItem;
    MenuItem1: TMenuItem;
    MenuItem2: TMenuItem;
    MenuEditCute: TMenuItem;
    MenuEditCopy: TMenuItem;
    MenuEditPaste: TMenuItem;
    MenuEditSelectAll: TMenuItem;
    MenuItem3: TMenuItem;
    N6: TMenuItem;
    N5: TMenuItem;
    MenuToolsEditorGutter: TMenuItem;
    N4: TMenuItem;
    N3: TMenuItem;
    MenuToolsSettings: TMenuItem;
    MenuTools: TMenuItem;
    MenuMain: TMainMenu;
    MenuFile: TMenuItem;
    MenuFileRefresh: TMenuItem;
    MenuFileExit: TMenuItem;
    MenuEdit: TMenuItem;
    MenuHelp: TMenuItem;
    MenuHelpAbout: TMenuItem;
    MenuFileExport: TMenuItem;
    MenuFileExportText: TMenuItem;
    MenuFileExportHTML: TMenuItem;
    N2: TMenuItem;
    N1: TMenuItem;
    PanelDBGrid: TPanel;
    SaveDialogMain: TSaveDialog;
    SelectDirectoryDialogMain: TSelectDirectoryDialog;
    SplitterDBGrid: TSplitter;
    StatusBarMain: TStatusBar;
    SynLuaSyn: TSynAnySyn;
    SynEditScript: TSynEdit;
    SynPythonSyn: TSynPythonSyn;
    SynXMLSyn: TSynXMLSyn;
    ToolBarMain: TToolBar;
    ToolButton1: TToolButton;
    ToolButton10: TToolButton;
    ToolButton11: TToolButton;
    ToolButton12: TToolButton;
    ToolButton13: TToolButton;
    ToolButton14: TToolButton;
    ToolButton16: TToolButton;
    ToolButton2: TToolButton;
    ToolButton3: TToolButton;
    ToolButton4: TToolButton;
    ToolButton8: TToolButton;
    procedure ActionEditCopyExecute(Sender: TObject);
    procedure ActionEditCopyScriptExecute(Sender: TObject);
    procedure ActionEditCutExecute(Sender: TObject);
    procedure ActionEditFilterExecute(Sender: TObject);
    procedure ActionEditFilterResetExecute(Sender: TObject);
    procedure ActionEditPasteExecute(Sender: TObject);
    procedure ActionEditSelectAllExecute(Sender: TObject);
    procedure ActionFileExitExecute(Sender: TObject);
    procedure ActionFileExportEachScriptToFileExecute(Sender: TObject);
    procedure ActionFileExportAllToTextExecute(Sender: TObject);
    procedure ActionFileExportAllToHTMLExecute(Sender: TObject);
    procedure ActionHelpAboutExecute(Sender: TObject);
    procedure ActionFileRefreshExecute(Sender: TObject);
    procedure ActionFileLoadExecute(Sender: TObject);
    procedure ActionToolsEditorFontSizeExecute(Sender: TObject);
    procedure ActionToolsEditorGutterExecute(Sender: TObject);
    procedure ActionToolsSettingsExecute(Sender: TObject);
    procedure DBGridDomoticzDrawColumnCell(Sender: TObject; const {%H-}Rect: TRect;
      {%H-}DataCol: Integer; {%H-}Column: TColumn; State: TGridDrawState);
    procedure FormCreate(Sender: TObject);
    procedure FormActivate(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var {%H-}CanClose: boolean);
    procedure FormShow(Sender: TObject);
   private
     FFirstTime: Boolean;               // Flag used to check if database to be loaded
     FSettingsFile: TIniFile;           // Inifile in the application folder
     FSettings: TSettingsData;          // Settings defined in a record
     procedure LoadSettings();
     procedure SaveSettings();
     procedure SetSynEditSyntaxHighligher;
   public
     // Used by the DBMgr
     procedure SetSynEditXMLStatement(AText: String);
 end;

var
 FormMain: TFormMain;

implementation

{$R *.lfm}
uses
 udbmgr;

procedure TFormMain.FormCreate(Sender: TObject);
begin
 FFirstTime := True;
 // Read the settings to be able to read the domoticz database
 LoadSettings();
end;

procedure TFormMain.FormShow(Sender: TObject);
begin
 Caption := VERSION;
 PanelDBGrid.Width := 240;
end;

procedure TFormMain.FormActivate(Sender: TObject);
begin
 // FormMain.Refresh;
 ActionFileRefreshExecute(Sender);
 SetSyneditSyntaxHighligher;
 FFirstTime := False;
end;

procedure TFormMain.FormCloseQuery(Sender: TObject; var CanClose: boolean);
begin
 DBMgr.CloseDatabase;
 SaveSettings;
end;

{REGION ACTIONS}
procedure TFormMain.ActionFileExitExecute(Sender: TObject);
begin
 Close;
end;

// Download the domoticz database using a shared folder on the rpi domoticz server
(*
The shared folder uses samba and is configured with:
sudo nano /etc/samba/smb.conf
[DoProDomoticz]
Comment = Raspberry Pi Domoticz Production folder
Path = /home/pi/domoticz
Browseable = yes
Writeable = Yes
only guest = no
create mask = 0777
directory mask = 0777
Public = yes
Guest ok = yes

Notes on the settings:
[DoProDomoticz]	= Name of the share
comment = The text is displayed as Comments in shares detail view
path = Specifies the folder that contains the files to be shared
browsable = yes|no - Set this share to be visible when running the net view command and also when you browse the network shares.
writable = yes|no - Allows user to add/modify files and folders in this share. By the default samba shares are readonly
guest ok = yes|no - Allows non authenticated users access the share

After configuration:
Ensure to set the directory mask for the domoticz folder to 0777:
sudo chmod 777 /home/pi/domoticz

Then restart samba:
sudo /etc/init.d/samba restart

And restart domoticz:
sudo service domoticz.sh restart

In windows connect to the shared folder using the Windows Explorer > Connect Network:
Path for drive Z: is \\domoticz-ip\DoProDomoticz
(DoProDomoticz is defined in the samba configuration)

After this prework, the domoticz database domoticz.db can be copied using CopyFile.
(simple solution using the standard lazarus fileutil functions).
*)
procedure TFormMain.ActionFileRefreshExecute(Sender: TObject);
const
  MSGTITLE = 'Refresh Database';
var
 bCopyDatabase: Boolean;
begin
 // Clear the dbgrid columns
 DBGridDomoticz.Columns.Clear;
 DBGridDomoticz.Visible := false;

 // Close the database first to avoid being locked
 DBMgr.CloseDatabase();

 bCopyDatabase := True;
 if FFirstTime and Not FSettings.RefreshAtStart then begin
   bCopyDatabase:= False;
 end;

 if bCopyDatabase or not FileExists(FSettings.DatabaseFile) then begin
  StatusBarMain.SimpleText := 'Please wait, copying the Domoticz database from the server...';
  StatusBarMain.Refresh;
  SynEditScript.Text := StatusBarMain.SimpleText;
  SynEditScript.Refresh;
  // If flag is set, then Download the database file from the domotic server
  // Example: z:\domoticz.db, domoticz.db
  if CopyFile(FSettings.SharedDrive+FSettings.DatabaseFile, FSettings.DatabaseFile) then begin
    StatusBarMain.SimpleText:='Domoticz database successfully copied from the Domoticz server.';
  end
  else begin
    StatusBarMain.SimpleText:='[Error] Domoticz database not copied from the Domoticz server.';
    ShowError(MSGTITLE,
      StatusBarMain.SimpleText
      +LineEnding
      +SysErrorMessage(GetLastOSError)
      +LineEnding+LineEnding
      +'Current database will be reloaded.');
    ActionFileLoadExecute(Sender);
    Exit;
  end;
 end;
 ActionFileLoadExecute(Sender);
end;

procedure TFormMain.ActionFileLoadExecute(Sender: TObject);
begin
  DBGridDomoticz.Visible := True;
  // Clear the dbgrid columns
 DBGridDomoticz.Columns.Clear;
 // Close the database first to avoid being locked
 DBMgr.CloseDatabase();
 if not FileExists(FSettings.DatabaseFile) then begin
    StatusBarMain.SimpleText:='[Error] Domoticz database not found.';
    ShowError('Load Database',
      StatusBarMain.SimpleText + LineEnding + 'Action: Refresh or copy the database to the application folder.');
 end;
 // Open the database, select the records, build the dbgrid
 // Hide the XMLStatement column
 if DBMgr.OpenDatabase(FSettings.DatabaseFile,
                       ExtractFilePath(Application.ExeName),
                       FSettings.SQLQuery) then begin
   DBGridDomoticz.Columns.Items[0].Width := 200;       // Name
   DBGridDomoticz.Columns.Items[1].Width := 100;       // Interpreter
   DBGridDomoticz.Columns.Items[2].Visible := False;   // XML-Statement
   DBGridDomoticz.Columns.Items[3].Width := 75;        // State
   DBGridDomoticz.Columns.Items[3].Field.Alignment := taLeftJustify;
   DBMgr.SetSQLQueryFilter(FSettings.Filter);
   StatusBarMain.SimpleText:='Domoticz database opened.';
 end
 else begin
   StatusBarMain.SimpleText:='[Error] Domoticz database not opened.';
   ShowError('Load Database',
     StatusBarMain.SimpleText + LineEnding+DBMgr.StatusMsg);
 end;
end;

procedure TFormMain.ActionToolsEditorFontSizeExecute(Sender: TObject);
var
  // InputResult: Boolean;
  InputString: String;
  FontSize: Integer;
begin
  InputString := IntToStr(SynEditScript.Font.Size);
  if InputQuery('Editor Font Size', 'Enter Size:', False, InputString) then begin
     try
       FontSize := StrToInt(InputString);
       SynEditScript.Font.Size := FontSize;
     except on E: Exception do begin
         ShowError('Editor Font Size', Format('Wrong font size entered: %s. Try again.',[InputString]));
       end;
     end;
  end;
end;

procedure TFormMain.ActionToolsEditorGutterExecute(Sender: TObject);
begin
  SynEditScript.Gutter.Visible := ActionToolsEditorGutter.Checked;
end;

procedure TFormMain.ActionEditCutExecute(Sender: TObject);
begin
    SynEditScript.CommandProcessor(TSynEditorCommand(ecCut), ' ', nil);
end;

procedure TFormMain.ActionEditCopyExecute(Sender: TObject);
begin
    SynEditScript.CommandProcessor(TSynEditorCommand(ecCopy), ' ', nil);
end;

procedure TFormMain.ActionEditPasteExecute(Sender: TObject);
begin
  SynEditScript.CommandProcessor(TSynEditorCommand(ecPaste), ' ', nil);
end;

procedure TFormMain.ActionEditSelectAllExecute(Sender: TObject);
begin
  SynEditScript.CommandProcessor(TSynEditorCommand(ecSelectAll), ' ', nil);
end;

// Copy the content of the selected script to the clipboard
procedure TFormMain.ActionEditCopyScriptExecute(Sender: TObject);
begin
 CopyToClipboard(SynEditScript.Text);
end;

procedure TFormMain.ActionEditFilterExecute(Sender: TObject);
begin
 // Ask for the query without masking (=hide text)
 if InputQuery(
      'Set Filter Condition',
      'Enter (Example:Interpreter="dzVents"; Empty field: filter reset):',
      False,
      FSettings.Filter) then begin
   DBMgr.SetSQLQueryFilter(FSettings.Filter);
   StatusBarMain.SimpleText:='Filter set: '  + FSettings.Filter;
  end;
end;

procedure TFormMain.ActionEditFilterResetExecute(Sender: TObject);
begin
  DBMgr.SetSQLQueryFilter('');
end;

// Export the listed scripts to text file
procedure TFormMain.ActionFileExportAllToTextExecute(Sender: TObject);
begin
  with SaveDialogMain do begin
    FileName := 'domoticzinternalscripts.txt';
    if Execute then begin
     DBGridDomoticz.BeginUpdate;
     DBMgr.ExportScriptsText(FileName);
     DBGridDomoticz.EndUpdate;
     if DBMgr.Status then begin
      StatusBarMain.SimpleText := Format(
        'Exported scripts to text file %s.',
        [FileName]);
      ShowInfo('Export Scripts Text File', StatusBarMain.SimpleText)
     end
     else begin
       StatusBarMain.SimpleText := Format(
         '[ERROR] Export scripts text file: %s.',
         [DBMgr.StatusMsg]);
       ShowError('Export Scripts Text File', StatusBarMain.SimpleText)
     end;
    end;
  end;
end;

procedure TFormMain.ActionFileExportAllToHTMLExecute(Sender: TObject);
begin
  with SaveDialogMain do begin
    FileName := 'domoticzinternalscripts.html';
    if Execute then begin
     DBGridDomoticz.BeginUpdate;
     DBMgr.ExportScriptsHTML(FileName);
     DBGridDomoticz.EndUpdate;
     if DBMgr.Status then begin
      StatusBarMain.SimpleText := Format(
        'Exported scripts to HTML file %s.',
        [FileName]);
      ShowInfo('Export Scripts HTML File', StatusBarMain.SimpleText)
     end
     else begin
       StatusBarMain.SimpleText := Format(
         '[ERROR] Export scripts HTML file: %s.',
         [DBMgr.StatusMsg]);
       ShowError('Export Scripts HTML File', StatusBarMain.SimpleText)
     end;
    end;
  end;
end;

procedure TFormMain.ActionFileExportEachScriptToFileExecute(Sender: TObject);
begin
 SelectDirectoryDialogMain.InitialDir := GetCurrentDir;
  if SelectDirectoryDialogMain.Execute then begin
   DBGridDomoticz.BeginUpdate;
   DBMgr.ExportScriptsTextEach(SelectDirectoryDialogMain.FileName);
   DBGridDomoticz.EndUpdate;
   if DBMgr.Status then begin
     StatusBarMain.SimpleText:='Scripts exported to folder ' + SelectDirectoryDialogMain.FileName;
     ShowInfo('Export Each Script Text File', StatusBarMain.SimpleText)
   end
   else begin
     StatusBarMain.SimpleText := DBMgr.StatusMsg;
     ShowError('Export Each Script Text File', StatusBarMain.SimpleText);
   end;
  end;
end;

procedure TFormMain.ActionToolsSettingsExecute(Sender: TObject);
var
 SQLQuery: String;
begin
 SqlQuery := FSettings.SQLQuery;
 Settings.SettingsData := FSettings;
 if Settings.ShowModal = mrOK then begin
   FSettings := Settings.SettingsData;
   // (Re)load the database in case the sql statement has changed
   if SQLQuery <> FSettings.SQLQuery then
      ActionFileLoadExecute(Sender);
 end;
end;

procedure TFormMain.ActionHelpAboutExecute(Sender: TObject);
begin
 ShowInfo(VERSION, COPYRIGHT);
end;
{ENDREGION ACTIONS}

{REGION DBGRID}
(*
  if a row in the dbgrid is selected, set the synedit syntax highligher.
*)
procedure TFormMain.DBGridDomoticzDrawColumnCell(Sender: TObject;
  const Rect: TRect; DataCol: Integer; Column: TColumn; State: TGridDrawState);
begin
  if gdSelected in State then begin
   Case DBGridDomoticz.DataSource.DataSet.FieldByName(FIELDINTERPRETER).AsString of
      INTERPRETERBLOCKLY: SynEditScript.Highlighter := SynXMLSyn;
      INTERPRETERDZVENTS: SynEditScript.Highlighter := SynLuaSyn;
      INTERPRETERPYTHON: SynEditScript.Highlighter := SynPythonSyn;
   end;
   StatusBarMain.SimpleText := Format('Selected: %s (%s)',
     [DBGridDomoticz.DataSource.DataSet.FieldByName(FIELDNAME).AsString,
      DBGridDomoticz.DataSource.DataSet.FieldByName(FIELDINTERPRETER).AsString]);
  end;
end;
{ENDREGION DBGRID}

{REGION SYNEDIT}
(*
  Used by the DBMgr
*)
procedure TFormMain.SetSynEditXMLStatement(AText: String);
begin
 AText := AText.Replace('><', '>'+LineEnding + '<');
 SynEditScript.Text := AText;               // DBGridDomoticz.Columns[2].Field.AsString;
end;
{ENDREGION SYNEDIT}

{REGION SETTINGS}
// Read the settings from the ini file located in the application folder
procedure TFormMain.LoadSettings();
begin
  FSettingsFile := TINIFile.Create(INIFILE);
  with FSettingsFile do begin
    with FSettings do begin
       SharedDrive    := ReadString(INISECTIONGENERAL,'shareddrive', SHAREDDRIVE);
       DatabaseFile   := ReadString(INISECTIONGENERAL,'databasefile', DATABASEFILE);
       RefreshAtStart := ReadBool(INISECTIONGENERAL,'refreshatstart', REFRESHATSTART);
       SQLQuery       := ReadString(INISECTIONGENERAL,'sqlquery', SQLQUERY);
       Filter         := ReadString(INISECTIONGENERAL,'filter', FILTER);
    end;
    // Editor
    ActionToolsEditorGutter.Checked := ReadBool(INISECTIONSYNEDIT,'gutter', False);
    SynEditScript.Gutter.Visible := ActionToolsEditorGutter.Checked;
    SynEditScript.Font.Size := ReadInteger(INISECTIONSYNEDIT,'fontsize', 9);
    SynLuaSyn.Constants.AddCommaText(ReadString(INISECTIONSYNEDIT, 'contants', ''));
    SynLuaSyn.Objects.AddCommaText(ReadString(INISECTIONSYNEDIT, 'objects', ''));
    SynLuaSyn.KeyWords.AddCommaText(ReadString(INISECTIONSYNEDIT, 'keywords', ''));
  end;
end;

// Save the settings to the ini file located in the application folder
procedure TFormMain.SaveSettings();
begin
  with FSettingsFile do begin
    with FSettings do begin
      WriteString(INISECTIONGENERAL,'shareddrive',SharedDrive);
      WriteString(INISECTIONGENERAL,'databasefile',DatabaseFile);
      WriteBool(INISECTIONGENERAL,'refreshatstart',RefreshAtStart);
      WriteString(INISECTIONGENERAL,'sqlquery',SQLQuery);
      WriteString(INISECTIONGENERAL,'filter',Filter);
    end;
    WriteBool(INISECTIONSYNEDIT,'gutter', ActionToolsEditorGutter.Checked);
    WriteInteger(INISECTIONSYNEDIT,'fontsize', SynEditScript.Font.Size);
  end;
end;

// Set syntax highlighter as (kind of) dzVents highlighter
procedure TFormMain.SetSyneditSyntaxHighligher;
begin
  SynLuaSyn.Constants.AddCommaText('DATA,DEVICES,HTTPRESPONSES,LOGGING,TIMER');
  SynLuaSyn.Objects.AddCommaText('LOCAL');
  SynLuaSyn.KeyWords.AddCommaText('ACTIVE,CUSTOMEVENTS,DATA,DEVICES,DUMP,ELSE,END,EXECUTE,FALSE,FILTER,FIND,FOR,FOREACH,FORMAT,FUNCTION,GROUPS,HELPERS,HTTPRESPONSES,IF,IN,LEVEL,LOG,LOGGING,MARKER,NOTIFY,ON,REDUCE,RETURN,SCENES,SECURITY,SHELLCOMMANDRESPONSES,SWITCHOFF,SWITCHON,SYSTEM,THEN,TIME,TIMER,TRUE,VARIABLES');
end;
{ENDREGION SETTINGS}

end.

